packages/query-core/src/queryObserver.ts

```diff
@@ -582,6 +592,11 @@ export class QueryObserver<
       isRefetchError: isError && hasData,
       isStale: isStale(query, options),
       refetch: this.refetch,
+      promise: hasData
+        ? fulfilledThenable(data as TData)
+        : status === 'error'
+          ? rejectedThenable(error)
+          : neverResolvedPromise,
```

Comment:
I do feel like this should be [a deferred](https://github.com/trpc/trpc/blob/f1efc8b29c562f4cc88add27aefaeb6a104b67ac/packages/server/src/unstable-core-do-not-import/stream/utils/createDeferred.ts#L2-L8) on the `QueryObserver` instead, but since none of the functions that transform results (`select`, etc) are stable, I don't know if that'd work well and the promise would still be recreated with every result... 

but there's probably ab better way of doing this, possibly one that doesn't need the `Thenable`-type at all...